numba_cuda/numba/cuda/core/event.py

```diff
@@ -0,0 +1,252 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: BSD-2-Clause
+
+"""
+The ``numba.cuda.core.event`` module provides a simple event system for applications
+to register callbacks to listen to specific compiler events.
+
+The following events are built in:
+
+- ``"numba:compile"`` is broadcast when a dispatcher is compiling. Events of
+  this kind have ``data`` defined to be a ``dict`` with the following
+  key-values:
+
+  - ``"dispatcher"``: the dispatcher object that is compiling.
+  - ``"args"``: the argument types.
+  - ``"return_type"``: the return type.
+
+- ``"numba:compiler_lock"`` is broadcast when the internal compiler-lock is
+  acquired. This is mostly used internally to measure time spent with the lock
+  acquired.
+
+- ``"numba:llvm_lock"`` is broadcast when the internal LLVM-lock is acquired.
+  This is used internally to measure time spent with the lock acquired.
+
+- ``"numba:run_pass"`` is broadcast when a compiler pass is running.
+
+    - ``"name"``: pass name.
+    - ``"qualname"``: qualified name of the function being compiled.
+    - ``"module"``: module name of the function being compiled.
+    - ``"flags"``: compilation flags.
+    - ``"args"``: argument types.
+    - ``"return_type"`` return type.
+
+Applications can register callbacks that are listening for specific events using
+``register(kind: str, listener: Listener)``, where ``listener`` is an instance
+of ``Listener`` that defines custom actions on occurrence of the specific event.
+"""
+
+import enum
+from contextlib import contextmanager, ExitStack
+from collections import defaultdict
+
+
+class EventStatus(enum.Enum):
+    """Status of an event."""
+
+    START = enum.auto()
+    END = enum.auto()
+
+
+# Builtin event kinds.
+_builtin_kinds = frozenset(
+    [
+        "numba:compiler_lock",
+        "numba:compile",
+        "numba:llvm_lock",
+        "numba:run_pass",
```

Comment:
Since this is not coming from Numba itself anymore, it's probably best that we rename these:

```suggestion
        "numba-cuda:compiler_lock",
        "numba-cuda:compile",
        "numba-cuda:llvm_lock",
        "numba-cuda:run_pass",
```

This might have a "downstream" effect where we need to rename the usage of events in the rest of the Numba-CUDA codebase so they begin with `numba-cuda` as well